### ISSUE TITLE
tensor.chunk returns wrong number of chunks

### ISSUE THREAD

**Issue by elimohl** [pytorch/pytorch#9382]
## Issue description
Number of chunks generated by `torch.chunk` doesn't match `chunks` parameter.

## Code example
```
>>> x = torch.arange(0, 11)
>>> x.chunk(5)
(tensor([ 0.,  1.,  2.]),
 tensor([ 3.,  4.,  5.]),
 tensor([ 6.,  7.,  8.]),
 tensor([  9.,  10.]))
>>> len(x.chunk(chunks=5))
4
```

## System Info
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Debian GNU/Linux unstable (sid)
GCC version: (Debian 7.3.0-24) 7.3.0
CMake version: version 3.11.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: GPU 0: GeForce GTX 950M
Nvidia driver version: 390.67
cuDNN version: Probably one of the following:
/usr/local/cuda/lib64/libcudnn.so.7.1.4
/usr/local/cuda/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.14.5)
[pip3] torch (0.4.0)
[pip3] torchvision (0.2.1)
[conda] Could not collect

pytorch installed by pip3

cc @ezyang @gchanan @zou3519 @mruberry @SsnL

**Comment by elimohl** [comment]
@gchanan `numpy.split` is equivalent to `pytorch.split` not to `pytorch.chunk`.
But Numpy can split array of size that is not divisible by the number of chunks with `numpy.array_split`: https://docs.scipy.org/doc/numpy/reference/generated/numpy.array_split.html

```
>>> import numpy as np
>>> x = np.arange(12)
>>> np.array_split(x, 5) 
[array([0, 1, 2]),
 array([3, 4, 5]),
 array([6, 7]),
 array([8, 9]),
 array([10, 11])]
```

**Comment by gchanan** [comment]
Numpy.split isn't equivalent to either, but it's closer to torch.chunk (the parameters are number of splits/chunks, not sizes).  Thanks for the pointer to array_split, we should use those semantics if the sizes aren't divisible.

**Comment by yf225** [comment]
Related: https://github.com/pytorch/pytorch/issues/8686.

We will need zero-dim size support to implement `numpy.split` semantics properly, because it supports returning sections that have zero elements, which means we need support for `narrow` with zero length.

**Comment by ptrblck** [comment]
The related issue seems to be closed by now.
Are we good to go with this issue?
If so, I would like to match the behavior of `torch.chunk` to that of `np.array_split`, if that's OK.

**Comment by vadimkantorov** [comment]
Just stumbled on this. It is indeed super-confusing that the number of returned chunks does not match the passed chunks argument (at least the documentation should reflect this):
```
In [39]: len(torch.arange(184).chunk(18))
Out[39]: 17
```

**Comment by ezyang** [comment]
Some history: https://github.com/torch/torch7/issues/617

**Comment by gchanan** [comment]
Too bad that history doesn't explain why it's "by design".

**Comment by vakkov** [comment]
It's high time that this gets solved

**Comment by leockl** [comment]
Hi @gchanan and @yf225,

Similar to what @ptrblck had mentioned, would be great to have `torch.chunk` match the behavior of that of `np.array_split`.

```
arr1 = np.array([[1.,2,3], [4,5,6], [7,8,9]])
tsr1 = torch.tensor(arr1).cuda()

tsr_chuck = torch.chunk(tsr1, 
                        chunks = 4,
                        dim = 0)

arr_split = np.array_split(arr1,
                           indices_or_sections = 4,
                           axis = 0)
```
```
tsr_chuck
```
Output:
```
(tensor([[1., 2., 3.]], device='cuda:0', dtype=torch.float64),
 tensor([[4., 5., 6.]], device='cuda:0', dtype=torch.float64),
 tensor([[7., 8., 9.]], device='cuda:0', dtype=torch.float64))
```
```
arr_split
```
Output:
```
[array([[1., 2., 3.]]),
 array([[4., 5., 6.]]),
 array([[7., 8., 9.]]),
 array([], shape=(0, 3), dtype=float64)]
```
where `np.array_split` includes an empty array when the number of `indices_or_sections` specified is greater than the number of dimensions in the specified `axis`.

**This issue has been sitting here for awhile now.** Will this ever get fixed?

**Comment by yf225** [comment]
@mruberry Curious would this be part of the numpy compat work? Thanks Mike!

**Comment by mruberry** [comment]
Definitely a candidate @yf225, but I have to review further.

**Comment by kurtamohler** [comment]
I can work on this. But first, I want to address how to deal with deprecation. Since we want a function whose behavior matches `numpy.array_split`, I suggest that we create a new function called `torch.tensor_split`, and then we can deprecate the `torch.chunk` method. Does that plan sound good?

**Comment by mruberry** [comment]
Creating `torch.tensor_split` sounds like a good first step.

**Comment by zou3519** [comment]
From an API perspective `torch.tensor_split` sounds really close to `torch.split` so it can be confusing. But I'm not sure what could be beter

**Comment by zou3519** [comment]
It's also worth noting that our [split function](https://pytorch.org/docs/stable/generated/torch.split.html) is different from numpy's [split function](https://numpy.org/doc/stable/reference/generated/numpy.split.html) and that numpy's split and array_split functions are pretty similar in terms of functionality so they don't really have this problem.

**Comment by kurtamohler** [comment]
I agree it could be confusing. However, it does get us closer to numpy-like compatibility. For now, I'll name it `torch.tensor_split`, and I can easily change it before the PR gets merged, if a better name comes up.
